File: core/model/certificate.go
package model

import "time"

type PrivateKeyInfo struct {
	Alias        string `json:"alias"`
	KeyAlgorithm string `json:"key_algorithm"`
	KeySize      int    `json:"key_size"`
	Pem          string `json:"pem"`
}

type CSRInfo struct {
	CommonName     string   `json:"common_name"`
	Country        string   `json:"country"`
	Organization   string   `json:"organization"`
	EmailAddresses []string `json:"email_addresses"`
	Pem            string   `json:"pem"`
}

type CertificateGroup struct {
	Alias        string                  `json:"alias"`
	Requester    string                  `json:"requester"`
	Certificates []CertificateSimpleInfo `json:"certificates"`
}

type CertificateSimpleInfo struct {
	Domain             string    `json:"domain"`
	AltDomains         []string  `json:"alt_names"`
	NotBefore          time.Time `json:"not_before"`
	NotAfter           time.Time `json:"not_after"`
	IssuerOrganization string    `json:"issuer_organization"`
	Status             string    `json:"status"`
	Source             string    `json:"source"`
}

type CertificateInfo struct {
	// 证书信息
	Domain       string    `json:"domain"`
	AltDomains   []string  `json:"alt_names"`
	NotBefore    time.Time `json:"not_before"`
	NotAfter     time.Time `json:"not_after"`
	Country      string    `json:"country"`
	Organization string    `json:"organization"`
	KeyAlgorithm string    `json:"key_algorithm"`
	KeySize      int       `json:"key_size"`
	IsCA         bool      `json:"is_ca"`

	// 签发机构信息
	IssuerCN           string `json:"issuer_cn"`
	IssuerCountry      string `json:"issuer_country"`
	IssuerOrganization string `json:"issuer_organization"`

	// 证书代码（PEM格式）
	Pem string `json:"pem"`

	Source string `json:"source"`
}

type CreateGroupRequest struct {
	Alias              string `json:"alias"`             // 主体别名
	DomainName         string `json:"domain_name"`       // 域名/名称
	Email              string `json:"email"`             // 邮箱
	Organization       string `json:"organization"`      // 公司/组织
	OrganizationalUnit string `json:"organization_unit"` // 部门
	Country            string `json:"country"`           // 国家
	Province           string `json:"province"`          // 省份
	City               string `json:"city"`              // 城市
	KeyAlgorithm       string `json:"key_algorithm"`     // 秘钥算法
}

type SelfSignedRequest struct {
	Alias       string `json:"alias"`        // 用于查找目录和证书项目的主体别名
	ExpireUnit  string `json:"expire_unit"`  // 有效期单位：day 或 year
	ExpireValue int    `json:"expire_value"` // 有效期值
	AltDomains  string `json:"alt_domains"`  // 备用域名，以换行符分隔
	AltIPs      string `json:"alt_ips"`      // 备用IP地址，以换行符分隔
	IsCA        bool   `json:"is_ca"`        // 是否允许使用该证书签发下级证书
}

type GroupPkRequest struct {
	Alias string `json:"alias"`
}

type CertificateInfoRequest struct {
	Source string `json:"source"`
}

type DeleteGroupRequest struct {
	Alias string `json:"alias"`
}

type DeleteCertificateRequest struct {
	Source string `json:"source"`
}

type ImportCertificateRequest struct {
	Alias         string `json:"alias"`
	KeyType       int    `json:"key_type"`
	KeyContent    string `json:"key_content"`
	KeyPath       string `json:"key_path"`
	CaType        int    `json:"ca_type"`
	CaContent     string `json:"ca_content"`
	CaPath        string `json:"ca_path"`
	CsrType       int    `json:"csr_type"`
	CsrContent    string `json:"csr_content"`
	CsrPath       string `json:"csr_path"`
	CompleteChain bool   `json:"complete_chain"`
}

File: frontend/src/api/certificate.ts
import request from '@/helper/api-helper';

// Certificate related interfaces
export interface PrivateKeyInfo {
  alias: string;
  key_algorithm: string;
  key_size: number;
  pem: string;
}

export interface CSRInfo {
  common_name: string;
  country: string;
  organization: string;
  email_addresses: string[];
  pem: string;
}

export interface CertificateSimpleInfo {
  domain: string;
  alt_names: string[];
  not_before: string;
  not_after: string;
  issuer_organization: string;
  status: string;
  source: string;
}

export interface CertificateGroup {
  alias: string;
  requester: string;
  certificates: CertificateSimpleInfo[];
}

export interface CertificateInfo {
  // 证书信息
  domain: string;
  alt_names: string[];
  not_before: string;
  not_after: string;
  country: string;
  organization: string;
  key_algorithm: string;
  key_size: number;
  is_ca: boolean;

  // 签发机构信息
  issuer_cn: string;
  issuer_country: string;
  issuer_organization: string;

  // 证书代码（PEM格式）
  pem: string;
  source: string;
}

// Request interfaces
export interface CreateGroupRequest {
  alias: string;
  domain_name: string;
  email: string;
  organization: string;
  organization_unit: string;
  country: string;
  province: string;
  city: string;
  key_algorithm: string;
}

export interface SelfSignedRequest {
  alias: string;
  expire_unit: 'day' | 'year';
  expire_value: number;
  alt_domains: string;
  alt_ips: string;
  is_ca: boolean;
}

export interface ImportCertificateRequest {
  alias: string;
  key_type: number;
  key_content?: string;
  key_path?: string;
  ca_type: number;
  ca_content?: string;
  ca_path?: string;
  csr_type: number;
  csr_content?: string;
  csr_path?: string;
}

export interface GroupPkRequest {
  alias: string;
}

export interface CertificateInfoRequest {
  source: string;
}

export interface DeleteGroupRequest {
  alias: string;
}

export interface DeleteCertificateRequest {
  source: string;
}

// API functions
/**
 * 获取证书组列表
 * @param hostId - 主机ID
 */
export function getCertificateGroups(hostId: number) {
  return request.get<{ items: CertificateGroup[] }>(
    `/certificates/${hostId}/group`
  );
}

/**
 * 创建证书组
 * @param hostId - 主机ID
 * @param data - 证书组创建参数
 */
export function createCertificateGroup(
  hostId: number,
  data: CreateGroupRequest
) {
  return request.post(`/certificates/${hostId}/group`, data);
}

/**
 * 删除证书组
 * @param hostId - 主机ID
 * @param data - 删除参数
 */
export function deleteCertificateGroup(
  hostId: number,
  data: DeleteGroupRequest
) {
  return request.delete(`/certificates/${hostId}/group`, data);
}

/**
 * 获取私钥信息
 * @param hostId - 主机ID
 * @param alias - 证书组别名
 */
export function getPrivateKeyInfo(hostId: number, alias: string) {
  return request.get<PrivateKeyInfo>(`/certificates/${hostId}/group/key`, {
    alias,
  });
}

/**
 * 获取CSR信息
 * @param hostId - 主机ID
 * @param alias - 证书组别名
 */
export function getCSRInfo(hostId: number, alias: string) {
  return request.get<CSRInfo>(`/certificates/${hostId}/group/csr`, { alias });
}

/**
 * 获取证书详细信息
 * @param hostId - 主机ID
 * @param source - 证书文件路径
 */
export function getCertificateInfo(hostId: number, source: string) {
  return request.get<CertificateInfo>(`/certificates/${hostId}`, { source });
}

/**
 * 删除证书
 * @param hostId - 主机ID
 * @param data - 删除参数
 */
export function deleteCertificate(
  hostId: number,
  data: DeleteCertificateRequest
) {
  return request.delete(`/certificates/${hostId}`, data);
}

/**
 * 生成自签名证书
 * @param hostId - 主机ID
 * @param data - 自签名证书参数
 */
export function generateSelfSignedCertificate(
  hostId: number,
  data: SelfSignedRequest
) {
  return request.post(`/certificates/${hostId}/sign/self`, data);
}

/**
 * 补齐证书链
 * @param hostId - 主机ID
 * @param data - 证书信息参数
 */
export function completeCertificateChain(
  hostId: number,
  data: CertificateInfoRequest
) {
  return request.post(`/certificates/${hostId}/complete`, data);
}

/**
 * 导入证书
 * @param hostId - 主机ID
 * @param formData - 表单数据
 */
export function importCertificate(hostId: number, formData: FormData) {
  return request.post(`/certificates/${hostId}/import`, formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
}
